src/pstl_offload.cpp

```diff
@@ -113,43 +126,223 @@ __original_msize(void* __user_ptr)
+    {
+        assert(__is_ptr_page_aligned(__ptr));
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        [[maybe_unused]] auto __ret = __this._M_map->emplace(__ptr, __ptr_desc{std::move(__device_ptr), __size});
+        assert(__ret.second); // the pointer must be unique
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<__ptr_desc>
+    __unregister_ptr(__large_aligned_ptrs_map& __this, void* __ptr)
+    {
+        // only page-aligned can be registered, so they may not be in the map
+        if (!__is_ptr_page_aligned(__ptr))
+        {
+            return std::nullopt;
+        }
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        if (!__this._M_map)
+        {
+            // ctor of static object not yet run, so it can't be our pointer
+            return std::nullopt;
+        }
+        auto __iter = __this._M_map->find(__ptr);
+        if (__iter == __this._M_map->end())
+        {
+            return std::nullopt;
+        }
+        __ptr_desc __header = std::move(__iter->second);
+        __this._M_map->erase(__iter);
+        return __header;
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<std::size_t>
+    __get_size(__large_aligned_ptrs_map& __this, void* __ptr)
+    {
+        // only page-aligned can be registered
+        if (!__is_ptr_page_aligned(__ptr))
+        {
+            return std::nullopt;
+        }
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        if (!__this._M_map)
+        {
+            // ctor of static object not yet run, so it can't be our pointer
+            return std::nullopt;
+        }
+        auto __iter = __this._M_map->find(__ptr);
+        if (__iter == __this._M_map->end())
+        {
+            return std::nullopt;
+        }
+        return __iter->second._M_requested_number_of_bytes;
+    }
+};
+
+static __large_aligned_ptrs_map __large_aligned_ptrs;
+
+void
+__block_header::__free()
+{
+    _M_uniq_const = 0;
+    sycl::context __context = _M_device.__get_context();
+    void* __original_pointer = _M_original_pointer;
+    this->~__block_header();
+    sycl::free(__original_pointer, __context);
+}
+
+#if __linux__
+
 static void
 __internal_free(void* __user_ptr)
 {
     if (__user_ptr != nullptr)
     {
-        __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+        if (__block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+            __same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+        {
+            __header->__free();
+            return;
+        }
 
-        if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+        if (std::optional<__large_aligned_ptrs_map::__ptr_desc>
+                __desc = __large_aligned_ptrs_map::__unregister_ptr(__large_aligned_ptrs, __user_ptr))
```

Comment:
I would suggest to rewrite `std::optional` checks here and in other places to initialization statement as you have above to decouple initialization statement and bool expression. Whether to use `operator bool` or `has_value` is up to you. I would use `has_value` for the sake of readability but this is a very slight preference.

For example for this code snippet I would do something like:
```suggestion
        if (std::optional<__large_aligned_ptrs_map::__ptr_desc>
                __desc = __large_aligned_ptrs_map::__unregister_ptr(__large_aligned_ptrs, __user_ptr); __desc.has_value())
```